File: src/spoj/dynamic_programming/GNYR09F.c
// http://www.spoj.com/problems/GNYR09F/

#include <stdio.h>

#define TRUE 1

typedef
struct {
    long endsWith_0;
    long endsWith_1;
    int isMemoized;
} MemoryCell;

#define MAX 100

MemoryCell MEMOIZED[MAX + 1][MAX + 1];

void calculate(int bitStrLen, int adjBitsCount) {
    if(bitStrLen == 1) {
        MEMOIZED[bitStrLen][adjBitsCount].endsWith_0 = 1;
        MEMOIZED[bitStrLen][adjBitsCount].endsWith_1 = 1;
        MEMOIZED[bitStrLen][adjBitsCount].isMemoized = TRUE;
        return;
    }
    
    if(adjBitsCount == 0) {
        if(!MEMOIZED[bitStrLen - 1][adjBitsCount].isMemoized) {
            calculate(bitStrLen - 1, adjBitsCount);
        }
        
        MEMOIZED[bitStrLen][adjBitsCount].endsWith_0 = MEMOIZED[bitStrLen - 1][adjBitsCount].endsWith_0 + MEMOIZED[bitStrLen - 1][adjBitsCount].endsWith_1;
        MEMOIZED[bitStrLen][adjBitsCount].endsWith_1 = MEMOIZED[bitStrLen - 1][adjBitsCount].endsWith_0;
        MEMOIZED[bitStrLen][adjBitsCount].isMemoized = TRUE;
        return;
    }
    
    if(adjBitsCount == bitStrLen - 1) {
        MEMOIZED[bitStrLen][adjBitsCount].endsWith_0 = 0;
        MEMOIZED[bitStrLen][adjBitsCount].endsWith_1 = 1;
        MEMOIZED[bitStrLen][adjBitsCount].isMemoized = TRUE;
        return;
    }
    
    if(!MEMOIZED[bitStrLen - 1][adjBitsCount - 1].isMemoized) {
        calculate(bitStrLen - 1, adjBitsCount - 1);
    }
    
    if(!MEMOIZED[bitStrLen - 1][adjBitsCount].isMemoized) {
        calculate(bitStrLen - 1, adjBitsCount);
    }
    
    // To get sequence of n bits, where (AdjBC == k), we have to:
    //
    // 1) Append '1' to:
    // - either to sequence of (n - 1) bits, where (AdjBC == (k - 1)), and ends with '1'
    // - or to sequence of (n - 1) bits, where (AdjBC == k), and ends with '0'
    //
    // 2) Append '0' to sequence of (n - 1) bits, where (AdjBC == k), which might ens either with '0', or with '1'
    
    // 1)
    MEMOIZED[bitStrLen][adjBitsCount].endsWith_1 = MEMOIZED[bitStrLen - 1][adjBitsCount - 1].endsWith_1 + MEMOIZED[bitStrLen - 1][adjBitsCount].endsWith_0;
    
    // 2)
    MEMOIZED[bitStrLen][adjBitsCount].endsWith_0 = MEMOIZED[bitStrLen - 1][adjBitsCount].endsWith_1 + MEMOIZED[bitStrLen - 1][adjBitsCount].endsWith_0;
    
    MEMOIZED[bitStrLen][adjBitsCount].isMemoized = TRUE;
}

long solve(int bitStrLen, int adjBitsCount) {
    calculate(bitStrLen, adjBitsCount);
    return MEMOIZED[bitStrLen][adjBitsCount].endsWith_0 + MEMOIZED[bitStrLen][adjBitsCount].endsWith_1;
}

int main() {
    /*
    printf("%ld\n", solve(5, 2));
    printf("%ld\n", solve(20, 8));
    printf("%ld\n", solve(30, 17));
    printf("%ld\n", solve(40, 24));
    printf("%ld\n", solve(50, 37));
    printf("%ld\n", solve(60, 52));
    printf("%ld\n", solve(70, 59));
    printf("%ld\n", solve(80, 73));
    printf("%ld\n", solve(90, 84));
    printf("%ld\n", solve(100, 90));
    */
    
    int testCases;
    int testCaseNum;
    int bitStrLen;
    int adjBitsCount;
    
    scanf("%d", &testCases);
    
    while(testCases--) {
        scanf("%d %d %d", &testCaseNum, &bitStrLen, &adjBitsCount);
        printf("%d %ld\n", testCaseNum, solve(bitStrLen, adjBitsCount));
    }
    
    return 0;
}

File: src/spoj/dynamic_programming/GNYR09F.java
package spoj.dynamic_programming;

// http://www.spoj.com/problems/GNYR09F/

/**
 *
 *
 * Input:
 * 5 2
 * 20 8
 * 30 17
 * 40 24
 * 50 37
 * 60 52
 * 70 59
 * 80 73
 * 90 84
 * 100 90
 *
 * Output:
 * 6
 * 63426
 * 1861225
 * 168212501
 * 44874764
 * 160916
 * 22937308
 * 99167
 * 15476
 * 23076518
 */

public class GNYR09F {

	public static void main(String... args) {
		System.out.println(solve(5, 2));
		System.out.println(solve(20, 8));
		System.out.println(solve(30, 17));
		System.out.println(solve(40, 24));
		System.out.println(solve(50, 37));
		System.out.println(solve(60, 52));
		System.out.println(solve(70, 59));
		System.out.println(solve(80, 73));
		System.out.println(solve(90, 84));
		System.out.println(solve(100, 90));
	}

	static long solve(int bitStringLen, int adjBitsCount) {
		Result r = f(bitStringLen, adjBitsCount);
		return r.e0 + r.e1;
	}

	static final int MAX_BIT_STRING_LEN = 100;

	static Result[][] MEMOIZED = new Result[MAX_BIT_STRING_LEN + 1][MAX_BIT_STRING_LEN + 1];

	static Result f(int bitStringLen, int adjBitsCount) {

		if ((adjBitsCount >= bitStringLen) || (adjBitsCount < 0)) {
			Result impossible = new Result();
			impossible.e0 = 0L;
			impossible.e1 = 0L;
			return impossible;
		}

		if (MEMOIZED[bitStringLen][adjBitsCount] != null) {
			return MEMOIZED[bitStringLen][adjBitsCount];
		}

		if (bitStringLen == 1) {
			Result smallestBitStr = new Result();
			smallestBitStr.e0 = 1L;
			smallestBitStr.e1 = 1L;
			MEMOIZED[bitStringLen][adjBitsCount] = smallestBitStr;
			return MEMOIZED[bitStringLen][adjBitsCount];
		}

		Result a = f(bitStringLen - 1, adjBitsCount - 1);
		Result b = f(bitStringLen - 1, adjBitsCount);

		Result result = new Result();
		result.e1 = a.e1 + b.e0;
		result.e0 = b.e0 + b.e1;

		MEMOIZED[bitStringLen][adjBitsCount] = result;
		return MEMOIZED[bitStringLen][adjBitsCount];
	}

	static class Result {
		// count of bit strings which ends with '1'
		Long e1;
		// count of bit strings which ends with '0'
		Long e0;
	}
}